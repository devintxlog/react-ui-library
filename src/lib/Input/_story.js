import React from 'react'
import { storiesOf } from '@storybook/react';
import Input from './'

storiesOf('Input', module)
.add('default', () => (
  <Input></Input>
))
.add('password', () => (
  <Input
    type='password'
  ></Input>
))

storiesOf('Input Select', module)
.add('default', () => (
  <Input
    type='select'
  ></Input>
))
.add('with options', () => (
  <Input
    type='select'
    defaultValue='default'
  >
    <option value="default" disabled>Choose an option</option>
    <option>Option 1</option>
  </Input>
))
