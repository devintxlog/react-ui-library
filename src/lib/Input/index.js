import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

//import css
import styles from './styles.module.scss'

//import validator
import validateString from './../../validators'

// logic behind all the different types of inputs
class Input extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      value: '',
      entered: false,
      error: false,
      infoText: null
    }
  }

  handleOnChange = (e) => {
    const val = e.target.value

    //validate if the user has entered the input once before
    if (this.state.entered && this.props.validate) {
      this.validate(val)
    }
    
    //update the state and fire the onChange function if one was passed in
    this.setState({value: val})
    this.props.onChange(val)  
  }

  handleOnBlur = () => {
    const val = this.state.value

    //only run validation if validate is set to true
    if (this.props.validate) {
      //set entered to true
      this.setState({entered: true})
      //run validation
      this.validate(val)
    }

    //fire the blur event that was passed in 
    this.props.onBlur(val)
  }

  validate = (val) => {
    const res = validateString(val, this.props.type)
    const isValid = res.valid
    const message = res.message
    const newValue = res.value

    //if it is not valid then set the error state
    if (!isValid) {
      this.setState({
        error: true,
        infoText: message
      })
    } else {
      this.setState({
        error: false,
        infoText: null
      })
      //fire the onValid function with 
      this.props.onValid(newValue)
    }
  }

  render(){
    //var to handle the html type
    let type = null
    //var to store the html element we are going to display
    let element = null

    //dynamically handle css classnames for inputs
    let inputClassNames = classNames({
      [styles.input]: true,
      [this.props.className]: this.props.className,
      [styles.error]: this.props.error || this.state.error
    })

    //dynamically handle css classnames for selects
    let selectClassNames = classNames({
      [styles.selectWrapper]: true,
      [this.props.className]: this.props.className,
      [styles.error]: this.props.error || this.state.error
    })

    //dynamically handle css classnames for info text
    let textClassNames = classNames({
      [styles.infoText]: true,
      [styles.error]: this.props.error || this.state.error
    })

    //logic for the different html input types
    switch (this.props.type) {
      case `password`:
        type = `password`
        break
      default:
        type = `text`
        break
    }

    //make the default html element an input
    element =  <input
      id={this.props.idForLabel}
      style={this.props.inlineStyles}
      className={inputClassNames}
      placeholder={this.props.placeholder}
      defaultValue={this.props.defaultValue}
      type={type}
      onChange={this.handleOnChange}
      onBlur={this.handleOnBlur}
    />

    //if type is select then make the element a select tag
    if (this.props.type === 'select') {
      element = <div
        className={selectClassNames} 
        style={this.props.inlineStyles} 
      >
        <select
          id={this.props.idForLabel}
          className={styles.select}
          defaultValue={this.props.defaultValue}
          onChange={this.handleOnChange}
        >
          {this.props.children}
        </select>
      </div>
    }

    return (
      <div>
        {element}
        <p className={textClassNames}>{this.props.infoText || this.state.infoText}</p>
      </div>
    )
  }
}

//rules for props being passed in
Input.propTypes = {
  idForLabel: PropTypes.string,
  className: PropTypes.string,
  inlineStyles: PropTypes.object,
  placeholder: PropTypes.string,
  infoText: PropTypes.string,
  password: PropTypes.bool,
  error: PropTypes.bool,
  defaultValue: PropTypes.string,
  validate: PropTypes.bool,
  type: PropTypes.oneOf(['text', 'email', 'password', 'select']).isRequired,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onValid: PropTypes.func
}

// Specifies the default values for props:
Input.defaultProps = {
  placeholder: ``,
  infoText: ``,
  idForLabel: ``,
  inlineStyles: {},
  password: false,
  error: false,
  defaultValue: '',
  validate: true,
  type: 'text',
  onChange: () => {},
  onBlur: () => {},
  onValid: () => {}
}

export default Input