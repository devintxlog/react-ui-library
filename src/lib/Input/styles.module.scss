//vars
$fontSize: 16px;
$lineHeight: 16px;
$colorBorder: #656565;

//base form element
%element {
  width: 100%;
  color: var(--colorFormInputFontColor);
  border-radius: 3px;
  border: 1px solid $colorBorder;
  transition: border .3s;
  box-shadow: inset 0 0 0 0 transparent;
  outline: 0;

  &:focus {
    border: 1px solid var(--colorPrimaryBase);
  }

  &.error {
    border: 1px solid red;
  }
}

.input {
  @extend %element;
  font-size: $fontSize;
  line-height: $lineHeight;
  padding: 5px 10px;

  //Style Placeholder
  &::-webkit-input-placeholder {
    color: var(--colorFormInputPlaceholder);
  }

  &:-moz-placeholder { /* Firefox 18- */
    color: var(--colorFormInputPlaceholder);
  }

  &::-moz-placeholder { /* Firefox 19+ */
    color: var(--colorFormInputPlaceholder);
  }

  &:-ms-input-placeholder {
    color: var(--colorFormInputPlaceholder);
  }
}

.select {
  background: transparent;
  border: 0;
  outline: 0;
  width: 100%;
  color: var(--colorFormInputFontColor);
  font-size: $fontSize;
  line-height: 20px;
  padding: 6px 10px;
  user-select: none;
  height: 32px;
  -webkit-appearance: none;
}

.selectWrapper {
  @extend %element;
  position: relative;
  overflow: hidden;
  background-color: white;

  &, > * {
    cursor: pointer;
  }

  &:after {
    content: '';
    width: 0;
    position: absolute;
    right: 7px;
    top: 50%;
    margin-top: -3px;
    border-width: 6px 4px;
    border-style: solid;
    pointer-events: none;
    border-color: $colorBorder transparent transparent transparent;
  }
}

.infoText {
  //keep the elements space even when it is blank
  display: inline-block;
  font-size: 12px;
  line-height: 16px;
  margin: 0;
  padding-left: 6px;

  &.error {
    color: red;
  }
}

