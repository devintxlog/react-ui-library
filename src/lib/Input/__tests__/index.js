import React from 'react'

import {  shallow } from 'enzyme'
import Input from '..'

describe('Input', () => {
  it('renders correctly', () => {
      const wrapper = shallow( < Input / > )
      expect(wrapper).toMatchSnapshot()
    }),
    describe('when password prop is true', () => {
      it('renders correctly', () => {
        const wrapper = shallow( <
          Input password = {
            true
          }
          />
        )
        expect(wrapper).toMatchSnapshot()
      })
    }),
    describe('when idForLabel prop is passed in', () => {
      it('renders correctly', () => {
        const wrapper = shallow( <
          Input idForLabel = {
            `exampleID`
          }
          />
        )
        expect(wrapper).toMatchSnapshot()
      })
    }),
    describe('when placeholder prop is passed in', () => {
      it('renders correctly', () => {
        const wrapper = shallow( <
          Input placeholder = {
            `Example Text`
          }
          />
        )
        expect(wrapper).toMatchSnapshot()
      })
    }),
    describe('when infoText prop is passed in', () => {
      it('renders correctly', () => {
        const wrapper = shallow( <
          Input infoText = {
            `Example Text`
          }
          />
        )
        expect(wrapper).toMatchSnapshot()
      })
    }),
    describe('when error prop is true', () => {
      it('renders correctly', () => {
        const wrapper = shallow( <
          Input error = {
            true
          }
          />
        )
        expect(wrapper).toMatchSnapshot()
      })
    }),
    describe('when value prop is passed in', () => {
      it('renders correctly', () => {
        const wrapper = shallow( <
          Input value = {
            `Example Text`
          }
          />
        )
        expect(wrapper).toMatchSnapshot()
      })
    }),
    describe('when handleOnChange is called', () => {
      it('calls the correct function and state is correct', () => {
        const e = {
          target: {
            value: 'test'
          }
        }
        const onChangeMock = jest.fn()
        const wrapper = shallow( <
          Input onChange = {
            onChangeMock
          }
          />
        )
        expect(wrapper.instance().handleOnChange(e)).toEqual(undefined)
        expect(wrapper.state().value).toEqual('test')
        expect(onChangeMock).toHaveBeenCalledTimes(1)
      })
    }),
    describe('when handleOnBlur is called', () => {
      it('calls the correct function', () => {
        const onBlurMock = jest.fn()
        const wrapper = shallow( <
          Input onBlur = {
            onBlurMock
          }
          />
        )
        expect(wrapper.instance().handleOnBlur()).toEqual(undefined)
        expect(onBlurMock).toHaveBeenCalledTimes(1)
      })
    })
})


//checkbox testing

describe('checkbox', () => {
  it('renders correctly', () => {
      const wrapper = shallow( <
        Input type = 'checkbox'
        idForLabel = {
          `exampleID`
        }
        />
      )
      expect(wrapper).toMatchSnapshot()
    }),
    describe('when handleOnChange is called', () => {
      it('calls the correct function and state is correct', () => {
        const e = {
          target: {
            value: 'test'
          }
        }
        const onChangeMock = jest.fn()
        const wrapper = shallow( <
          Input type = 'checkbox'
          idForLabel = {
            `exampleID`
          }
          onChange = {
            onChangeMock
          }
          />
        )
        expect(wrapper.instance().handleOnChange(e)).toEqual(undefined)
        expect(wrapper.state().value).toEqual('test')
        expect(onChangeMock).toHaveBeenCalledTimes(1)
      })
    }),
    describe('when handleOnBlur is called', () => {
      it('calls the correct function', () => {
        const onBlurMock = jest.fn()
        const wrapper = shallow( <
          Input type = 'checkbox'
          idForLabel = {
            `exampleID`
          }
          onBlur = {
            onBlurMock
          }
          />
        )
        expect(wrapper.instance().handleOnBlur()).toEqual(undefined)
        expect(onBlurMock).toHaveBeenCalledTimes(1)
      })
    })
})


//Radio testing

describe('radio', () => {
  it('renders correctly', () => {
      const wrapper = shallow( <
        Input type = 'radio'
        idForLabel = {
          `exampleID`
        }
        />
      )
      expect(wrapper).toMatchSnapshot()
    }),
    describe('when handleOnChange is called', () => {
      it('calls the correct function and state is correct', () => {
        const e = {
          target: {
            value: 'test'
          }
        }
        const onChangeMock = jest.fn()
        const wrapper = shallow( <
          Input type = 'radio'
          idForLabel = {
            `exampleID`
          }
          onChange = {
            onChangeMock
          }
          />
        )
        expect(wrapper.instance().handleOnChange(e)).toEqual(undefined)
        expect(wrapper.state().value).toEqual('test')
        expect(onChangeMock).toHaveBeenCalledTimes(1)
      })
    }),
    describe('when handleOnBlur is called', () => {
      it('calls the correct function', () => {
        const onBlurMock = jest.fn()
        const wrapper = shallow( <
          Input type = 'radio'
          idForLabel = {
            `exampleID`
          }
          onBlur = {
            onBlurMock
          }
          />
        )
        expect(wrapper.instance().handleOnBlur()).toEqual(undefined)
        expect(onBlurMock).toHaveBeenCalledTimes(1)
      })
    })

})
