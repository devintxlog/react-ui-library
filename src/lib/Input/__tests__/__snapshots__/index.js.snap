// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input checkbox renders correctly 1`] = `
<div>
  <Checkbox
    id="exampleID"
    labelText="None Provided"
    onBlur={[Function]}
    onChange={[Function]}
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input radio renders correctly 1`] = `
<div>
  <RadioButton
    id="exampleID"
    labelText="None Provided"
    onBlur={[Function]}
    onChange={[Function]}
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input select renders correctly when disabled is true 1`] = `
<div>
  <Select
    defaultValue=""
    disabled={true}
    error={false}
    id=""
    onBlur={[Function]}
    onChange={[Function]}
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input select renders correctly when type is select 1`] = `
<div>
  <Select
    defaultValue=""
    error={false}
    id=""
    onBlur={[Function]}
    onChange={[Function]}
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input textArea renders correctly 1`] = `
<div>
  <TextArea
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input textArea renders correctly when disabled is true 1`] = `
<div>
  <TextArea
    defaultValue=""
    disabled={true}
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input textArea renders correctly when type is textArea 1`] = `
<div>
  <TextArea
    defaultValue=""
    error={false}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input when disabled is passed in renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    disabled={true}
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input when error prop is true renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText error"
  />
</div>
`;

exports[`Input when idForLabel prop is passed in renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id="exampleID"
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input when infoText prop is passed in renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText"
  >
    Example Text
  </p>
</div>
`;

exports[`Input when placeholder prop is passed in renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Example Text"
    type="text"
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input when required is passed in renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input when type is password renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="password"
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input when validate is passed in renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText"
  />
</div>
`;

exports[`Input when value prop is passed in renders correctly 1`] = `
<div>
  <TextInput
    defaultValue=""
    id=""
    inlineStyles={Object {}}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    type="text"
  />
  <p
    className="infoText"
  />
</div>
`;
